### DISCUSSION TITLE
Naming for an attribute to enable "light dismiss for dialog"

### DISCUSSION THREAD

**Discussion by mfreed7** [openui/open-ui#950] [Polls]
There's a desire to add an attribute on `<dialog>` that adds "light dismiss" behavior to the dialog. The proposal is to add that via an attribute:

`<dialog attribute_name_here> I'm a light dismiss dialog </dialog>`

Adding `attribute_name_here` to the <dialog> makes the dialog get closed (as if dialog.close() were called) when the user ["light dismisses"](https://developer.mozilla.org/en-US/docs/Web/API/Popover_API/Using#auto_state_and_light_dismiss) it, e.g. by clicking outside the bounds of the dialog. 

Note: "Esc" key handling is separate to this light dismiss behaviour.

**Comment by matthewp** [comment]
Shouldn't this be the default? Why is it not?

**Reply by keithamus** [reply]
It cannot be the default for web compatibility reasons. See https://github.com/whatwg/html/issues/9373#issuecomment-1576940150

**Reply by matthewp** [reply]
Yeah, totally understand that it can't be the default *now*, I'm more curious why it didn't work this way from the beginning. If you know the answer or have any background links I'd love to understand more.

**Reply by domenic** [reply]
Most dialogs on your computer cannot be closed by clicking outside of them. Checking the ones I have quick access to:

- Firefox about dialog
- Microsoft Outlook add account dialog
- Notepad Save As dialog
- Visual Studio Code "There are currently no updates available" dialog

**Comment by matthewp** [comment]
I'm not familiar with the "light dismiss" phrase. Is there any background or links on what that means?

**Reply by keithamus** [reply]
Please see https://developer.mozilla.org/en-US/docs/Web/API/Popover_API/Using#auto_state_and_light_dismiss

**Reply by lukewarlow** [reply]
The below section explains how to achieve what this attribute would do in your own code. To give a concrete example of what "light dismiss" means in this context.

https://web.dev/articles/building/a-dialog-component#adding_light_dismiss

**Reply by matthewp** [reply]
Interesting, so it sounds like this an old term that I guess I've just somehow never encountered?

**Reply by matthewp** [reply]
I only bring this up as 1 data point, I would be confused seeing this attribute name. But if it's common knowledge then dismiss this data point.

**Comment by jamiebuilds** [comment]
In design systems I've always seen clicks like this referred to as "outside clicks" for all sorts of popover elements. I feel like that makes more sense than "light". Also if there are ever any future options here, I'd rather they be exposed as a list:

```html
<dialog closetrigger="outside delay blur etc">
```

**Reply by WickyNilliams** [reply]
"outside" seems too focused on the pointer/click aspect, when Esc is also a trigger for dismiss

**Reply by domenic** [reply]
The Esc trigger is separate from this discussion: https://github.com/whatwg/html/issues/9373#issuecomment-1720157787

**Reply by WickyNilliams** [reply]
Oh I thought light dismiss behavior covered both. The MDN linked article in the OP says it encompasses both. 

> The popover can be "light dismissed" — this means that you can hide the popover by clicking outside it or pressing the Esc key. 

Why is Esc considered "heavy" in the other issue? If you're a keyboard user it's no different to a click outside. If this is the case, I think the entire terminology is confusing and should be reconsidered

If they're separate, does that mean you can have one behavior but not the other?  Under what circumstances would you want that? I thought the intention to provide a uniform/reliable baseline pattern for dismissing? If it's not consistent it becomes frustrating, because the user never knows what to do

Edit: I see now that esc is already handled by the dialog. Though I still think the heavy/light terminology is confusing. Muddied further by MDN describing both as light dismiss

**Reply by domenic** [reply]
> The MDN linked article in the OP says it encompasses both.

It's unfortunate MDN is using this "light dismiss" term in a way that doesn't match the spec, and it's even more unfortunate that we still are using the "dismiss" term in the spec at all; see https://github.com/openui/open-ui/discussions/950#discussioncomment-7592785.

> If they're separate, does that mean you can have one behavior but not the other?

Yes. For example, `<dialog>` today has Esc key close behavior, and no click-outside close behavior.

>  Under what circumstances would you want that?

Generally, more-transient dialogs, where an accidental click outside closing the dialog is not a problem, want the click-outside behavior. Whereas all dialogs want the Esc key behavior, which is a very explicit close request.

See also this comment https://github.com/openui/open-ui/discussions/950#discussioncomment-7592839 where I list various dialogs which have this behavior (Esc key closes, click-outside does not) on my computer.

**Reply by lukewarlow** [reply]
Wrt MDN https://github.com/mdn/content/issues/30309 I raised an issue to sort out the fact it refers to escape as light dismiss

**Reply by karlhorky** [reply]
I like `closetrigger` and exposing a list 👍 

However, as @domenic mentioned in [the comment below](https://github.com/openui/open-ui/discussions/950#discussioncomment-7592862), `::backdrop` is the official name of the pseudo-element, so would be probably a better replacement for subjective words like "outside":

```html
<dialog closetrigger="backdrop delay blur etc">
```

**Reply by zandermax** [reply]
I'm a big fan of this, since it uses existing terminology with no potential spelling pitfalls, there's no ambiguity to what it refers to - i.e. these things will trigger `.close()` - and using a list to allow for several ways to close it in a single property.

If this is a list that can contain multiple values, what about using `closetriggers` ? Is there a paradigm we are attempting to follow by using the singular?

What about using `closeon` ? I am aware this is close to `onclose`, but because it is flipped would it be more intuitive? I am actually liking it a bit better because `closetrigger` suggests that the list must be comprehensive, i.e. these are *all* of the ways this dialog can be closed (does it need to include `submit` for example?), whereas `closeon` does not suggest the same.

```html
<dialog closeon="backdrop delay blur etc">
```

**Reply by domenic** [reply]
Some further discussion of this list idea in https://github.com/whatwg/html/issues/9373#issuecomment-1822014942

**Reply by lukewarlow** [reply]
For me closeon is far too close to onclose and will lead to confusion especially when teaching to new developers.

I like the list idea closes="" or closetrigger(s) seem nice?

**Comment by brunoprietog** [comment]
close-on-click-outside?

**Comment by patrickgwsmith** [comment]
In Cocoa this property is called “transient”. See https://developer.apple.com/documentation/appkit/nspopover/behavior/transient

```html
<dialog transient> Tap the background to close me. </dialog>
```

I like @jamiebuilds idea of it being extensible so this attribute could also support a string value.

```html
<dialog transient="outside blur etc"> Focus anything else to close me. </dialog>
```

(Btw I’m not a fan of the term “light close” because it sounds like it could be referring to the light DOM.)

**Comment by CyberAP** [comment]
`dismiss="light"` to better follow the language of the popover spec?

**Reply by simevidas** [reply]
Or just boolean `dismiss`, since it’s a simple opt in.

**Comment by WebReflection** [comment]
`<dialog disposable>`

**Comment by mtgentry** [comment]
Soft-dismissal?

**Comment by aarjithn** [comment]
onblur-close

Or

close-onblur

**Comment by dfkaye** [comment]
`<dialog dismissible>`

**Reply by GrimLink** [reply]
This makes the most sense for the attribute name 👍

**Reply by simevidas** [reply]
Maybe the `-ible` suffix isn’t needed. Keeping it short sounds good. These are just keywords that developers will get used to anyway.

**Reply by keithamus** [reply]
My two concerns around `dismissible`:

 - Words with doubled letters can be tricky to spell (especially given the word contains 3 s letters in total). Similarly `ible`/`able` can be something folks trip up on. Given the presence of an incorrectly spelled attribute will silently fail, I can imagine a world with lots of `dismissable`, `dissmissable`, `dissmisable`, `dissmisble`... attributes that do nothing.
 - Dialogs are always dismissible with the Esc key. It may be surprising to someone that removing `dismissible` from `<dialog dismissible>` does not make it non-dismissible.

**Comment by nuxodin** [comment]
Could you make it the default behavior, but clicking on the backdrop prevents it from closing?
Then this could work:
```css
.dialog::backdrop {
    pointer-events: none;
}
```

**Reply by lukewarlow** [reply]
How would it be the default if clicking on the backdrop didn't cause this?

**Reply by nuxodin** [reply]
I mean when clicking outside the dialogue element **including backdrop**, closes the dialog

**Comment by lukewarlow** [comment]
Android flips the default (click outside closes by default), but they have a setCanceledOnTouchOutside() https://developer.android.com/reference/android/app/Dialog#setCanceledOnTouchOutside(boolean)

I think it's interesting that they use the cancel rather than close terminology here.

@mfreed7 am I correct in thinking this behaviour would trigger a cancel even like escape and back gesture do currently?

If so is it worth considering using the cancel term instead of close? (Not overly fussed just thought it was interesting)

**Reply by WebReflection** [reply]
I like `cancelable` too :+1:

**Reply by keithamus** [reply]
@WebReflection if you like this would you kindly upvote this comment, as it'll make it easier to collect votes.

**Reply by WebReflection** [reply]
forgot to ... done

**Reply by domenic** [reply]
For `CloseWatcher`, the two methods are `close()` (goes straight to closing) and `requestClose()` (also fires `cancel`, giving you a chance to abort the close request).

**Comment by mfreed7** [comment]
> @mfreed7 am I correct in thinking this behaviour would trigger a cancel even like escape and back gesture do currently?
> 
> If so is it worth considering using the cancel term instead of close? (Not overly fussed just thought it was interesting)

I was thinking this new behavior would trigger a close, not a cancel, but I don't have strong opinions. @domenic you have a better understanding of close vs. cancel, and stronger opinions about the naming. What do you think?

**Reply by domenic** [reply]
I think generally cancels are useful: clicking outside a dialog should give authors the opportunity to pop up an "are you sure want to close this dialog? you have unsaved changes" sub-dialog, or other UI.

However, right now dialog doesn't have any method of triggering a cancel besides close requests. E.g. there's no `requestClose()` like `CloseWatcher` has. Saying that you can only get a cancel via close requests or via click-outside is a bit strange, at that point. It might be good to solve that problem first.

**Reply by lukewarlow** [reply]
This was something we recently discussed in openUI wrt invokers. Am I to understand that you'd be open to a cancel (name tbd) function added to dialogs that fires cancel and close events? To allow a dialog cancel button to correctly reflect that in the events.

**Reply by domenic** [reply]
Yes, and I think it should be called requestClose() to match CloseWatcher.

**Comment by mfreed7** [comment]
So with 47 votes at this point, the winner with 27% is... `<dialog other>`! 😵‍💫  Clearly there's no obvious good name here. (I strongly suspect that if `lightdismiss` were included in the list, it likely would have won, because https://github.com/openui/open-ui/issues/834#issuecomment-1787725119.)

Looks like perhaps I need to collect some of the many suggestions made in comments here and run another survey? Some of the interesting ones include:
- `<dialog transient>`
- `<dialog disposable>`
- `<dialog cancelable>` (UK/US debatable spelling here)

Note that I skipped ones that have "dismiss" in them due to https://github.com/openui/open-ui/issues/834#issuecomment-1788292050, and those that try to specify the particular mechanisms to light dismiss (e.g. click vs. blur) due to https://github.com/openui/open-ui/issues/834#issuecomment-1786147470.

@domenic you have strong opinions that "dismiss" should *not* be in the name. Do you have issues with unrelated words like `transient`? Personally I like that one the best, since it doesn't imply *how* the dialog can be dismissed or what happens when it is dismissed, but *does* imply the dialog is easier to get rid of, which is really what light dismiss means in some sense.

**Reply by WebReflection** [reply]
FWIW I’d be ok with transient and where there’s some prior case to reason about it’s easier for everyone else to reason about it too 👍

**Reply by domenic** [reply]
My request is that "close" is in the name, since it triggers the `close` event and behaves like the `close()` method. If you want to use another word in such a name, like `transientclose`, that seems reasonable.

It's true that I'd like to completely stamp out this usage of the term "dismiss", which is just confusing. We already have close as a perfectly good verb for dialogs. Not to mention cancel, and toggle, and beforetoggle. We don't need yet another term. I think it'd be helpful if we were careful in how we talked about it, e.g. instead of your above sentence, we'd say

> Personally I like that one the best, since it doesn't imply how the dialog can be <ins>closed</ins> or what happens when it is <ins>closed</ins>, but does imply the dialog is easier to get rid of, which is really what light <ins>close</ins> means in some sense.

Maybe another good step would be to remove the usage of the term "dismiss" in the HTML spec text.

**Reply by WebReflection** [reply]
Is closable on the table?

**Comment by domenic** [comment]
I wonder about explicitly using the word "backdrop", to match the `::backdrop` API that this represents? Remember, the behavior we're trying to sum up is "closes when clicking on the backdrop".

- `backdropclose=""`
- `closeviabackdrop=""`

?

**Reply by karlhorky** [reply]
or adding `backdrop` to the `closetrigger` list like @jamiebuilds suggestion in [the comment above](https://github.com/openui/open-ui/discussions/950#discussioncomment-7580932)

```html
<dialog closetrigger="backdrop delay blur etc">
```

**Reply by WickyNilliams** [reply]
I really like the idea of using "backdrop" as part of the vocabulary of this API. Not sure about it being in the attribute name though

**Comment by domenic** [comment]
Note that `close=light/explicit/implicit` is not really possible because it collides with the existing `close()` method. (Similar problem with the old `shadowroot=""` attribute proposal that collided with the `shadowRoot` property and needed to get renamed to `shadowrootmode=""`.)

**Comment by mfreed7** [comment]
FYI, this conversation seems to have moved to https://github.com/whatwg/html/issues/9373.

I guess we'll pick it up there. It's too bad it's happening in two places.

**Comment by mfreed7** [comment]
Based on that conversation, I've created two more polls, to replace this one:

https://github.com/openui/open-ui/discussions/960
https://github.com/openui/open-ui/discussions/961
